### ISSUE TITLE
[MSHARED-1153] FilteringUtils makes incorrect assumption about filesystem

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-filtering#267]
**[James Nord](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jnord_cbs)** opened **[MSHARED-1153](https://issues.apache.org/jira/browse/MSHARED-1153?redirect=false)** and commented

when copying files using filtering during a project build the build can fail with the following exception due to an incorrect assumption in [FilteringUtils.java#L474-L489](https://github.com/apache/maven-filtering/blob/456b03a9510f1e64e457f393a4e04cbeeebd9538/src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java#L474-L489)

The filesystem in use may present a posix view (ie it is a mounted filesystem) but itself may not support setting posix attributes.

This is common when using docker on windows and mounting windows directory into the container.

this then causes the following exception :

```
Caused by: org.apache.maven.shared.filtering.MavenFilteringException: copying /home/ath-user/code/src/main/resources/ath-container/Dockerfile to /home/ath-user/code/target/classes/ath-container/Dockerfile failed with FileSystemException: /home/ath-user/code/target/classes/ath-container/Dockerfile: Operation not permitted
    at org.apache.maven.shared.filtering.DefaultMavenFileFilter.copyFile (DefaultMavenFileFilter.java:118)
    at org.apache.maven.shared.filtering.DefaultMavenResourcesFiltering.filterResources (DefaultMavenResourcesFiltering.java:277)
    at org.apache.maven.plugins.resources.ResourcesMojo.execute (ResourcesMojo.java:356)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:370)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.nio.file.FileSystemException: /home/ath-user/code/target/classes/ath-container/Dockerfile: Operation not permitted
    at sun.nio.fs.UnixException.translateToIOException (UnixException.java:100)
    at sun.nio.fs.UnixException.rethrowAsIOException (UnixException.java:111)
    at sun.nio.fs.UnixException.rethrowAsIOException (UnixException.java:116)
    at sun.nio.fs.UnixFileAttributeViews$Posix.setMode (UnixFileAttributeViews.java:254)
    at sun.nio.fs.UnixFileAttributeViews$Posix.setPermissions (UnixFileAttributeViews.java:276)
    at java.nio.file.Files.setPosixFilePermissions (Files.java:2080)
    at org.apache.maven.shared.filtering.FilteringUtils.copyFilePermissions (FilteringUtils.java:477)
    at org.apache.maven.shared.filtering.FilteringUtils.copyFile (FilteringUtils.java:457)
    at org.apache.maven.shared.filtering.DefaultMavenFileFilter.copyFile (DefaultMavenFileFilter.java:110)
    at org.apache.maven.shared.filtering.DefaultMavenResourcesFiltering.filterResources (DefaultMavenResourcesFiltering.java:277)
    at org.apache.maven.plugins.resources.ResourcesMojo.execute (ResourcesMojo.java:356)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:370)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

Given the code ignores setting attributes if it can not due to an UnsupportedOperation (e.g. the filesystem is not even posix capable) it should do the same (for the FileFileSystemException)
 
java 11.0.16.1 / WIndows 11 /  Docker Desktop 4.8.2 (79419).

Maven 3.8.6  / maven-resources-plugin 3.3.0

---

**Remote Links:**
- [GitHub Pull Request #116
](https://github.com/apache/maven-filtering/pull/116)

**Backported to:** [maven-filtering-3.4.1](https://github.com/apache/maven-filtering/milestone/22?closed=1)

**Comment by jira-importer** [comment]
**[James Nord](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jnord_cbs)** commented

2 years later and I am hitting this issue again (although possibly a different reproducer).

 

```
Caused by: java.nio.file.FileSystemException: /mnt/d/source/github/jenkinsci/acceptance-test-harness/target/classes/ath-container/Dockerfile: Operation not permitted
    at sun.nio.fs.UnixException.translateToIOException (UnixException.java:100)
    at sun.nio.fs.UnixException.rethrowAsIOException (UnixException.java:106)
    at sun.nio.fs.UnixException.rethrowAsIOException (UnixException.java:111)
    at sun.nio.fs.UnixFileAttributeViews$Posix.setMode (UnixFileAttributeViews.java:277)
    at sun.nio.fs.UnixFileAttributeViews$Posix.setPermissions (UnixFileAttributeViews.java:299)
    at java.nio.file.Files.setPosixFilePermissions (Files.java:2169)
    at org.apache.maven.shared.filtering.FilteringUtils.copyFilePermissions (FilteringUtils.java:410)
    at org.apache.maven.shared.filtering.FilteringUtils.copyFile (FilteringUtils.java:394) 
```

chmod reports

```
chmod: changing permissions of '/mnt/d/source/github/jenkinsci/acceptance-test-harness/target/classes/ath-container/Dockerfile': Operation not permitted 
```

`/mnt/d` is a mount of windows `D` drive in WSL.
The mount is then volume-mapped (`{}-v /mnt/d/:/mnt/d{`}) in WSL into a docker container.

I can run `chmod o-w /mnt/d/source/github/jenkinsci/acceptance-test-harness/target/classes/ath-container/Dockerfile` **outside** of the docker container.  this command returned an exit code of `0` but does not actually change the attributes (because its not a posix filesytem underneath).

So there is something about the container, or how docker desktop is mounting the mapped volume that triggers the issue.

Maven 3.9.9
Java  21.0.4, vendor: Ubuntu, runtime: /usr/lib/jvm/java-21-openjdk-amd64
maven-resources-plugin 3.2.2

**Comment by jira-importer** [comment]
**[James Nord](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jnord_cbs)** commented

I think adding the `FileSystemException` to the catch block [here](https://github.com/apache/maven-filtering/blob/bc81950fd60277af62c300b252a2d258e9d62e4a/src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java#L341C18-L348) would "fix" this and not make things unduley worse.
The code is already saying setting permissions is a best effort by performing a fallback, so this would just enhance the area where the fallback is allowed.

FTR: I also have a ticket opened with Docker to see if we can get to the bottom of why it is behaving so differently.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[James Nord](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jnord_cbs), you are right. The behavior violates the contract. Can you raise a PR for that?
Please also link the docker issue.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[James Nord](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jnord_cbs), please link the upstream issue.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [0e96f0113a8cd92616dd3251e5ed6cef18eafed7](https://gitbox.apache.org/repos/asf?p=maven-filtering.git&a=commit&h=0e96f0113a8cd92616dd3251e5ed6cef18eafed7) and with [2860039597ab60827a1b6347889aa112f285bf5f](https://gitbox.apache.org/repos/asf?p=maven-filtering.git&a=commit&h=2860039597ab60827a1b6347889aa112f285bf5f) for 3.x.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

You wrote here: https://issues.apache.org/jira/browse/MSHARED-1153?focusedCommentId=17879264&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17879264 that you have created an issue with upstream. It should be mentioned here.
